Deprecate use of type in reindex request body

cbuescher opened the pull request:
Types can be used both in the `source` and `dest` section of the body which will
be translated to search and index requests respectively. Adding a deprecation warning
for those cases and removing examples using more than one type in reindex since
support for this is going to be removed.

elasticmachine commented:
Pinging @elastic/es-search

mayya-sharipova reviewed (commented):
Thanks @cbuescher, looks good overall.
The main question I have is about the interpretation of "Deprecate use of type".  I imagined that for any reference to `type`, even though it is equal to `_doc`, we should issue a deprecation warning.  @jtibshirani  what do you think?

cbuescher commented on docs/reference/docs/reindex.asciidoc:
```diff
@@ -162,27 +161,8 @@ POST _reindex
-// CONSOLE
-// TEST[setup:twitter]
-// TEST[s/^/PUT blog\/post\/post1?refresh\n{"test": "foo"}\n/]
+`index` in `source` can be a list, allowing you to copy from
```
Good point, I I was probably deleting too much here. The two indices case should stay.

cbuescher commented:
>  I imagined that for any reference to `type`, even though it is equal to `_doc`, we should issue a deprecation warning.

I was going after something like I saw in RestGetAction where we check `!type.equals(MapperService.SINGLE_MAPPING_NAME)` before we log the deprecation. But maybe thats different there because "_doc" can still appear in the path. Here we can probably deprecate all uses because "type" is used in the request body.

jtibshirani commented:
I agree that we want to log a deprecation warning anytime that `type` appears in the request body. Even if `type` is set to `_doc`, we'll want to issue a warning because that key shouldn't be used at all.

I also wanted to note that you may be seeing search-related deprecation warnings when using `type` in the `source` section of a reindex request. This is because reindex can create a query on the `_type` field as part of its execution. I just merged #36802 to avoid this warning, so it might be good to pull in that commit.

cbuescher commented:
@jtibshirani @mayya-sharipova thanks for the feedback, I pushed an update.

jtibshirani commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java:
```diff
@@ -68,6 +72,7 @@
             }
             String[] types = extractStringArray(source, "type");
             if (types != null) {
+                deprecationLogger.deprecatedAndMaybeLog("reindex_with_types", TYPES_DEPRECATION_MESSAGE);
```
For me it'd be clearer to move these checks out of the parsing logic into `buildRequest`, where you can examine the parsed `ReindexRequest`. This is just a personal preference though, totally up to you if you'd like to change.

cbuescher replied:
I took a look at the alternative but I think I prefer it here

jtibshirani commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/RestReindexActionTests.java:
```diff
@@ -200,4 +209,41 @@ private RemoteInfo buildRemoteInfoHostTestCase(String hostInRest) throws IOExcep
+        FakeRestRequest.Builder requestBuilder = new FakeRestRequest.Builder(xContentRegistry())
+                .withMethod(Method.POST)
+                .withPath("/_reindex");
+        final StringBuilder requestContent = new StringBuilder();
```
Also just a personal preference, but for me it'd be easier to read if we used an `XContentBuilder` here to build up the source/ destination.

cbuescher replied:
I changed this to XContentBuilder.

jtibshirani commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/RestReindexActionTests.java:
```diff
@@ -200,4 +209,41 @@ private RemoteInfo buildRemoteInfoHostTestCase(String hostInRest) throws IOExcep
+        requestContent.append("  }");
+        requestContent.append("}");
+        requestBuilder.withContent(new BytesArray(requestContent.toString()), XContentType.JSON);
+        System.out.println(requestContent.toString());
```
Looks like a leftover debugging statement.

jtibshirani commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/RestReindexActionTests.java:
```diff
@@ -27,18 +27,27 @@
+
+    @Before
+    public void setUpAction() {
+        new RestReindexAction(Settings.EMPTY, controller());
```
Maybe we could update the other cases in this file to share this `action` instead of recreating one?

cbuescher replied:
sure

jtibshirani reviewed (commented):
Thanks @cbuescher, this is looking good to me! I think we'll also need to update `ReindexIT` to remove references to types.

mayya-sharipova reviewed (approved):
Thanks @cbuescher, LGTM

cbuescher commented:
@jtibshirani I think I adressed you latest round of comments, running the test now but I think this should be good to go now?

cbuescher commented:
@elasticmachine test this please

cbuescher commented:
@elasticmachine test this again please

cbuescher commented:
@elasticmachine test this please

cbuescher commented:
@elasticmachine test this please

Commit:
Deprecate use of type in reindex request body

Types can be used both in the source and dest section of the body which will
be translated to search and index requests respectively. Adding a deprecation warning
for those cases and removing examples using more than one type in reindex since
support for this is going to be removed.

cbuescher commented:
@elasticmachine test this please

cbuescher commented:
@elasticmachine test this please

Conky5 commented:
@elasticmachine test this please

cbuescher commented:
> Even if `type` is set to `_doc`, we'll want to issue a warning because that key shouldn't be used at all.

@jtibshirani I just that now the IndexRequest (used for the reindex destination) always contains a non-null type ("_doc" by default, this might be a recent change, I haven't seen related tests failing before). Also the SearchRequest used in the source section always prints the "type" field to xContent even if no types are set, so we will alway see a warning when going e.g. through the High Level Rest Client. I think we need to change the ReindexRequest xContent output slightly so that the "type" fields are not printed when they are not set (e.g. when coming from HLRest)

Commit:
Don't output empty of default type field in toXContent

cbuescher commented:
@jtibshirani thanks for the thumbs up but the last failures in the high level client tests uncovered what I think were some problems with the xContent output of the current ReindexRequest. Can you take a look at the recent commit and let me know if you agree with not outputting empty or default type fields in the ReindexRequest is serialized via toXContent? This would mean we still warn when empty/default types names are sent directly via REST, but we wouldn't send them outselves when using e.g. the HL client.

jtibshirani commented on server/src/main/java/org/elasticsearch/index/reindex/ReindexRequest.java:
```diff
@@ -294,15 +295,19 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
             builder.array("index", getSearchRequest().indices());
-            builder.array("type", getSearchRequest().types());
+            String[] types = getSearchRequest().types();
+            if (types != null && types.length > 0) {
```
Super small comment -- I don't think these null checks are necessary, since `SearchRequest` validates that the `types` array can't be null. Similarly, `IndexRequest` validates that the type is not null.

cbuescher replied:
> `IndexRequest` validates that the type is not null

Unfortunately this is done in `validate()` only, so its possible to set a `null` type via the ctors or the setter and then xContent serialization would break. For `SearchRequest` I think you are right and types cannot be `null` so I will remove the check there but leave it for IndexRequest.

jtibshirani commented:
@cbuescher thanks for double-checking, I agree with this approach to `ReindexRequest#toXContent`.

Commit:
iter

kcm commented:
This bit us on site today because the 7.x docs don't mention `type`, but it's still needed for reindexing from older multi-type indices into 7.x+. Even if the support for multiple types in `dest` is gone, won't we still want to keep the support for `type` in `source`? What's a workaround if not, for upgrading from pre-6 indices?

Can we at least re-document it for `source` for now?
